crates/telio-batcher/src/batcher.rs

```diff
@@ -0,0 +1,526 @@
+    async fn remove(&self, key: String);
+    async fn add(
+        &self,
+        key: String,
+        interval: std::time::Duration,
+        threshold: std::time::Duration,
+    ) -> std::result::Result<tokio::sync::watch::Receiver<()>, Error>;
+}
+
+struct State {
+    notifiers: HashMap<String, BatchEntry>,
+}
+
+pub struct Batcher {
+    task: Task<State>,
+}
+
+struct BatchEntry {
+    deadline: tokio::time::Instant,
+    interval: std::time::Duration,
+    threshold: std::time::Duration,
+    tx: tokio::sync::watch::Sender<()>,
+}
+
+#[async_trait]
+impl Runtime for State {
+    const NAME: &'static str = "Batcher";
+    type Err = ();
+
+    async fn wait_with_update<F>(&mut self, update: F) -> std::result::Result<(), Self::Err>
+    where
+        F: Future<Output = BoxAction<Self, std::result::Result<(), Self::Err>>> + Send,
+    {
+        telio_log_info!("****** Batcher wait_with_update");
+
+        let mut sleep_futures = {
+            let futures: FuturesUnordered<_> = self
+                .notifiers
+                .iter()
+                .map(|(key, entry)| {
+                    let deadline = entry.deadline;
+                    let key_clone = key.clone();
+                    telio_log_debug!("****** deadline => {:?}", deadline);
+                    async move {
+                        telio_log_debug!(
+                            "******  now befor sleep => {:?}",
+                            tokio::time::Instant::now()
+                        );
+                        tokio::time::sleep_until(deadline).await;
+                        telio_log_debug!(
+                            "******  now after sleep => {:?}",
+                            tokio::time::Instant::now()
+                        );
+                        key_clone
+                    }
+                })
+                .collect();
+
+            futures
+        };
+
+        telio_log_debug!(
+            "batcher before tokio::select! with {} futures",
+            sleep_futures.len()
+        );
+        tokio::select! {
+            update = update => {
+                telio_log_debug!("batcher inside tokio::select! @1");
+                return update(self).await;
+            }
+            Some(_) = sleep_futures.next() => {
+                let now = tokio::time::Instant::now();
+
+                let mut batched_count = 0;
+                let mut keys_to_delete = vec![];
+
+                {
+                    for (key, &mut ref mut entry) in self.notifiers.iter_mut() {
+                        let deadline_with_threshold = entry.deadline - entry.threshold;
+                        if now >= deadline_with_threshold {
+                            batched_count +=1;
+                            telio_log_debug!("batcher inside tokio::select! Batch!, will add new sleep with interval {:?} @ now {:?}", entry.interval, now);
+                            entry.deadline = now + entry.interval;
+
```

Comment:
It's apparent that maybe SessionKeeper could be responsible for all of the ICMP ECHO requests/pings that are being done by persistent_keepalive feature of WireGuard. Given that it looks like the case, it might be a much cleaner approach to:
- move batcher into session keeper
- restore repeated_actions, rename to batcher and modify to be able to work with thresholds.

This way modifications to the code would be kept at a minimum and no future-proofing would be there as it might be unecessary.